proto/api.go

```diff
@@ -79,6 +79,7 @@ type Request interface {
 	gogoproto.Message
 	// Header returns the request header.
 	Header() *RequestHeader
+	KVHeader() *KVRequestHeader
```

Comment:
Instead of KVHeader returning nil for structured requests, I'd rather split the interfaces as Pete described and have the senders that need KV fields (just DistSender?) type-assert the Request to a KVRequest. This is consistent with how a lot of the stdlib works (i.e. "this method takes an io.Reader. If it is also an io.ReadCloser it will be closed as well").